orttraining/orttraining/training_ops/cuda/communication/nccl_service.h

```diff
@@ -0,0 +1,239 @@
+    NcclTask scheduled_task;
+    scheduled_task.type = type;
+    scheduled_task.peers = peers;
+
+    for (auto& task : batch) {
+      if (!task.Compare(scheduled_task)) {
+        continue;
+      }
+      if (task.is_finished || task.is_enqueued) {
+        throw;
+      }
+
+      task.ptr = ptr;
+      task.size = size;
+      task.is_enqueued = true;
+      task.info = info;
+      return &task;
+    }
+
+    return nullptr;
+  };
+
+  bool IsAllTasksEqueued() {
+    for (auto& task : batch) {
+      if (!task.is_enqueued) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  bool IsAllTasksFinished() {
+    for (auto& task : batch) {
+      if (!task.is_finished) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  void ResetAllTasks () {
+    for (auto& task : batch) {
+      task.ptr = nullptr;
+      task.size = 0;
+      task.is_enqueued = false;
+      task.is_finished = false;
+    }
+  };
+
+  void Show(const std::string& prefix) const {
+    for (int i = 0; static_cast<size_t>(i) < batch.size(); ++i) {
+      std::string line = prefix;
+      auto& task = batch[i];
+      if (task.type == NcclTask::Type::SEND) {
+        line += "Send, ";
+      } else if (task.type == NcclTask::Type::RECV) {
+        line += "Recv, ";
+      }
+
+      for (int j = 0; static_cast<size_t>(j) < task.peers.size(); ++j) {
+        line += std::to_string(task.peers[j]);
+        if (static_cast<size_t>(j) != task.peers.size() - 1) {
+          line += ", ";
+        } else {
+          line += "\n";
+        }
+      }
+      std::cout << line;
+    }
+  }
+
+  std::vector<NcclTask> batch;
+};
+
+class NcclService final {
+ public:
+  static NcclService& GetInstance() {
+    static NcclService instance_;
+    return instance_;
+  };
+
+  void SubmitSendAndWait(void* buffer, size_t count, int peer);
+
+  void SubmitRecvAndWait(void* buffer, size_t count, int peer);
+
+  void StartPlan() {
+    if (is_planned_) {
+      throw;
+    }
+  };
+
+  void PlanStartNewGroup() {
+    group_status_.push_back(true); 
+    schedule_.push_back(NcclTaskGroup());
+  };
+  void PlanSend(const int dst) {
+    if (!group_status_.back()) {
+      throw;
+    }
+    
+    schedule_.back().PlanTask(NcclTask::Type::SEND, {dst});
+  };
+  void PlanRecv(const int src) {
+    if (!group_status_.back()) {
+      throw;
+    }
+    schedule_.back().PlanTask(NcclTask::Type::RECV, {src});
+  };
+  void PlanEndNewGroup() {
+    group_status_.back() = false;
+  }
+  void EndPlan() {
+    is_planned_ = true;
+  }
+
+  void Launch();
+
+  void Reset();
+
+  void Terminate();
+
+  void Show() const {
+    for (int i = 0; static_cast<size_t>(i) < schedule_.size(); ++i) {
+      std::cout << "Time " << i << std::endl;
+      schedule_[i].Show("  ");
+    }
+  }
+
+ private:
+  NcclService() = default;
+  ~NcclService() = default;
+  NcclService(const NcclService&) = delete;
+  NcclService& operator=(const NcclService&) = delete;
```

Comment:
I assume 'move' is also not allowed? Consider using ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE for these methods.